Added all single-file functionality to the lj script

MrBartWolf opened the pull request:
Added testing of the program to the lj script, replaced the scary Java.Applications error with a warning, and updated the README.

lam2mo commented:
@MrBartWolf Thanks! Will you also please change the `dotests.sh` script so that the Travis tests run and pass again?

lam2mo commented:
Sorry, that's not the problem. The problem is that you've removed the `test.sh` script, and that's what Travis is using for the tests (see `.travis.yml`). A quick fix would be just to re-add that script in the `tests` folder and update the Travis configuration. A more substantive fix would involve handling #50.

MrBartWolf commented:
The dotests.sh script will still fail since all of its scripts have been removed. I'm still working on rebuilding it, but it will likely take time. For now I will add the test.sh script for Travis.

lam2mo commented:
Note: also update `.travis.yml` to just call `dotests.sh` and make sure it's behaving the way Travis expects.

lam2mo commented:
Looks like `dotests.sh` no longer has execute permissions. Can you add those back in?

lam2mo commented:
Excellent, thanks.

@Zamua are you ok with merging all this in now?

lam2mo commented:
Actually, hang on. There seem to be some failures in the Travis logs even though for some reason it thinks it's passing. We should investigate that before we merge this in.

MrBartWolf commented:
The test that is failing is the primitive-numbers test that I'm fixing in another branch. As far as I can tell there is no other issue with the pull request. This request should be merged before the primitive-numbers one.

Zamua commented on dotests.sh:
```diff
@@ -25,64 +20,36 @@ test() {
         if $set_expected; then
-            cp $base\_compile.out $base\_compile.exp
-            cp $base\_run.out $base\_run.exp
-            cp $base\_test.out $base\_test.exp
```
Why remove these two lines?

MrBartWolf replied:
There are no run.* or test.* files anymore. The lj script is doing everything, and the output of it is getting put into one file.

Zamua commented on dotests.sh:
```diff
@@ -124,14 +91,6 @@ if [ $# == 0 ]; then
     exit
```
We should exit 1 if differences were detected. It is currently always exiting with 0 so travis will never fail

MrBartWolf replied:
It doesn't appear that the old dotests.sh script ever does exit 1 so I'm confused on what you mean.

Zamua replied:
the old .travis.yml script didn't run the `dotests.sh` script. It ran `./compile.sh <file> && ./test.sh`

If `./compile.sh` failed then it would return a non zero exit code here: https://github.com/JMU-CS/less-java/blob/master/compile.sh#L50

If `./test.sh` failed then it would return a non zero exit code here: https://github.com/JMU-CS/less-java/blob/master/test.sh#L17

Now that .travis.yml is not referencing these scripts and is instead referencing the `dotest.sh` script, the `./dotest.sh` script must return non zero exit codes on failure to notify travis that testing has failed.

lam2mo replied:
Yes, we discussed this at the meeting today and @MrBartWolf has instructions for wrapping this up.

lam2mo commented:
@MrBartWolf Please provide an appropriate summary message on this thread once this is ready to be merged. Use a bullet list to describe each of the changes made.

Note to whoever merges this PR once it's done: please squash with the aforementioned message.

MrBartWolf commented:
- Added testing of files to lj script
- Replaced error message of programs not having main with a warning, and stopped the lj script from running programs without a main function
- Rebuilt dotests.sh so it works with the lj script where all output is in a single file
- Reset all test outputs to match the new format.
- Set the travis file to call dotests.sh over the entire tests directory
- Added the exit 1 condition for any failing outputs in dotests.sh (travis will fail incorrect outputs)
- Fixed the primitive-numbers test and removed it from the failing directory
- Fixed incorrect io tests that didn't match the language and reset their output.
- Added .in files for any io tests that required them and were lacking them.
- Updated the README

Zamua commented:
Looks good to me.

All that's left is to rewrite the commit history a little to make it cleaner and easier to read. If you would like to do this yourself I recommend doing an interactive rebase.

**Below is an explanation of how to interactive rebase. Skip this if you are familiar with it.**

It will let you replay prior commits and will let you reorder, reword remove, edit, etc. I recommend rewording the top one, and "fixing up" (removing) the remaining. For the one you reword, I recommend pasting your entire summary comment into it.

The git commands will look something like this:
```
# 25 beause that's the number of commits in this PR
git rebase -i HEAD~25
...
# here vim is open and listing all of the selected commits.
# replace the word "pick" with "f" if you want to remove the commit (it will keep the changes, just not the commit message).
# Replace the word "pick" with "r" if you want to keep the commit message but you want to reword it.
# :wq the buffer and your rebase will be complete
# There will be instructions at the bottom of the vim window as well.
...
```

I also recommend checking out a new branch and experimenting first to ensure you understand what will happen and how to do it. If you'd rather not do this, then @lam2mo or I can do it ourselves before we merge in your PR.

lam2mo commented:
@MrBartWolf was encountering some strange issues with the rebase so I decided just to squash and merge from the web interface to avoid prolonging this PR any more. Everything seems fine so far.
